keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,445 @@
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.
+
+Existing logic:
+1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] This proposal requires a user to apply a `quarantine` taint on a node when the user has confirmed that this node is shutdown or in a non-recoerable state due to the hardware failure or broken OS.
+
+1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:
+  1.  Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.
+  1. new `quarantine` taint has NoSchedule effect, meaning it will not schedule on the shutdown node. For example, `quarantine=nodeshutdown:NoSchedule` or `quarantine=hardwarefailure:NoSchedule`.
+
+1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachement` to be deleted.
+
+1. This would trigger the deletion of the `volumeAttachement` objects. This would allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+1. When the `external-attacher` detects the `volumeAttachement` object is being deleted, it calls CSI driver's `ControllerUnpublishVolume`.
+
+### Handle the Return of Shutdown Node
+
+If the nodes that were previously shutdown with pods forcefully deleted come back, we need to make sure they are cleaned up before pods are scheduled on them again.
+
+[Proposed Change] Pod GC Controller automatically removes the `quarantine` taint. It checks to make sure no pods are attached to a Node, and no volume attachments on Node. If so, it removes the taint. This makes sure no new pods are scheduled on the node before the `quarantine` taint is removed.
```

Comment:
Do we need an option to disable automatic removing the `quarantine` taint, for example, to examine the root cause of node failure?

Reply:
Changed the KEP to require the user to manually remove the `quarantine` taint. Pod GC Controller can be enhanced in the future to automatically remove it.